### PR TITLE
Local alloc elemwise2

### PR DESCRIPTION
**Pull request by daemonmaker**

### CODE REVIEW

File: theano/sandbox/cuda/opt.py [hunk 1]
```diff
@@ -1814,6 +1814,12 @@ def split_huge_add_or_mul(node):
 optdb.register('gpu_inplace_elemwise_opt', gpu_inplace_elemwise_optimizer, 75,
                'fast_run', 'inplace', 'gpu_inplace')
 
+tensor.opt.register_specialize_device(tensor.opt.local_shape_to_shape_i)
```
Comment (nouiz) [inline theano/sandbox/cuda/opt.py]: I think both opt should be registered only in register_opt. If you do that, do it work?

File: theano/sandbox/cuda/tests/test_opt.py [hunk 2]
```diff
@@ -86,6 +87,176 @@ def test_gpualloc():
     assert numpy.any([isinstance(x.op, cuda.GpuAlloc) for x in l])
 
 
+class Test_local_elemwise_alloc(unittest.TestCase):
```
Comment (nouiz) [inline theano/sandbox/cuda/tests/test_opt.py]: Duplicating the test isn't good. Can you inherit from the original test in tensor/tests/test_opt.py and override the assert fct? Was there other change? You can see an example in cuda/tests/test_basic_ops.py with the class T_Join_and_Split

File: theano/sandbox/cuda/opt.py [hunk 3]
```diff
@@ -1814,6 +1814,14 @@ def split_huge_add_or_mul(node):
...
                'fast_run', 'inplace', 'gpu_inplace')
 
+register_opt()(tensor.opt.local_remove_useless_assert)
+
+register_opt()(tensor.opt.local_shape_to_shape_i)
+gpu_elemwise_alloc = gof.local_optimizer([GpuElemwise])(
+    tensor.opt.local_elemwise_alloc_op(GpuElemwise, GpuAlloc, GpuDimShuffle)
+)
+register_opt()(gpu_elemwise_alloc)
+tensor.opt.register_specialize_device(gpu_elemwise_alloc)
```
Comment (nouiz) [inline theano/sandbox/cuda/opt.py]: Do you still need to register in specialize_device? If it work without this, I prefer that you remove it.

File: theano/sandbox/cuda/tests/test_opt.py [hunk 4]
```diff
@@ -86,6 +88,50 @@ def test_gpualloc():
...
+        #self.vec = tensor.vector('vec', dtype=dtype)
+        #self.mat = tensor.matrix('mat', dtype=dtype)
+        #self.tens = tensor.tensor3('tens', dtype=dtype)
+
+        #self.alloc_wo_dep = basic_ops.gpu_alloc(self.vec, 2, 2)
+        #self.alloc_w_dep = basic_ops.gpu_alloc(self.vec, *self.mat.shape)
+
+        self.alloc_wo_dep = basic_ops.gpu_alloc(self.vec, 2, 2)
+        self.alloc_w_dep = basic_ops.gpu_alloc(self.vec, *self.mat.shape)
+        self.alloc_w_dep_tens = basic_ops.gpu_alloc(
```
Comment (nouiz) [inline theano/sandbox/cuda/tests/test_opt.py]: Why this one? I think it build the same graph as alloc_w_dep.

Comment (nouiz) [inline theano/sandbox/cuda/tests/test_opt.py]: ok, it was tested. This depend on tens and not mat and do not use all shape of tens. So this test a different use case.

Comment (daemonmaker) [inline theano/sandbox/cuda/tests/test_opt.py]: Just to be clear, you agree that this is correct, right?

Comment (nouiz) [inline theano/sandbox/cuda/tests/test_opt.py]: yes. The PR is good. I ran all tests. thanks I merge.

**Pull request merged**